src/main/java/org/example/resolver/generator/SchemaGenerator.java

```diff
@@ -0,0 +1,171 @@
+                importWithCall(dependency, writer, outputDirectoryPath, clazz);
+                importDone.add(dependency);
+            } else if (!importDone.contains(dependency)) {
+                importWithoutCall(dependency, writer, clazz);
+                importDone.add(dependency);
+            }
+        }
+
+        // Imports for Fields
+        for (Class<?> dependency : fields) {
+            if (ProtobufUtils.isPrimitiveType(dependency)) {
+                continue;
+            }
+            if (dependency.equals(Object.class) || !importDone.contains(dependency)) {
+                importWithoutCall(dependency, writer, clazz);
+                importDone.add(dependency);
+            } else if (!isFileGenerated.contains(dependency)) {
+                importWithCall(dependency, writer, outputDirectoryPath, clazz);
+                importDone.add(dependency);
+            }
+        }
+    }
+
+    private void writeMessage(BufferedWriter writer, Class<?> clazz, Set<Class<?>> interfaces, Set<Class<?>> superClass) throws IOException {
+
+        int tagNumber = FileCreator.schemaDependency(writer, clazz, interfaces, superClass);
+
+        // for every Field
+        Field[] fields = clazz.getDeclaredFields();
+        for (Field field : fields) {
+            Class<?> fieldType = field.getType();
+
+            if (ProtobufUtils.isPrimitiveListType(fieldType)) { // List Type
+                tagNumber = ListProcessor.listScan(field, tagNumber, writer);
+            } else if (ProtobufUtils.isPrimitiveMapType(fieldType)) {   // Map type
+                tagNumber = MapProcessor.mapScan(field, tagNumber, writer);
+            } else if (fieldType.isArray()) {  // Array Type
+                tagNumber = ArrayProcessor.arrayScan(field, tagNumber, writer);
+            } else if (fieldType.isEnum()) {    // Enum Type
+                tagNumber = EnumProcessor.enumScan(field, fieldType, tagNumber, writer);
+            } else if (fieldType.equals(Object.class)) {    // Object Type
+                tagNumber = ObjectProcessor.objectScan(field, writer, tagNumber, 1);
+            } else if (ProtobufUtils.isPrimitiveType(fieldType)) {
+
+                Class<?> protobufType = ProtobufUtils.getProtoPrimitiveType(fieldType);
+                writer.write("  " + protobufType.getSimpleName() + " " + field.getName() + " = " + (tagNumber++) + ";");
+                writer.newLine();
+
+            } else if (FileAnalyzer.checkNonPrimitive(fieldType)) {
+
+                writer.write("  " + fieldType.getSimpleName() + " " + field.getName() + " = " + (tagNumber++) + ";");
+                writer.newLine();
+            }
+        }
+        writer.write("}");
+        writer.newLine();
+        writer.close();
+    }
+
+
+    private void writeProtobufSchema(Class<?> clazz, String outputDirectoryPath) {
+
+        if (isFileGenerated.contains(clazz)){
+            return;
+        }
+        try {
+            isFileGenerated.add(clazz);
```

Comment:
The contains and add operations are atomic but both of them together are not. Two threads can check "clazz" doesn't exists before adding it and can try to add later.
Can use synchronized block or ConcurrentHashMap putIfAbsent() Method.

Reply:
synchronized added to make operations atomic.